lib/std/channels.nim

```diff
@@ -0,0 +1,703 @@
+
+  chan.tail = newTail
+  signal(chan.notEmptyCond)
+  result = true
+
+proc channel_recv_spsc(chan: ChannelRaw, data: pointer, size: int32, nonBlocking: bool): bool =
+  # Cannot be inline due to function table
+  channel_recv_mpsc(chan, data, size, nonBlocking)
+
+proc channel_close_spsc(chan: ChannelRaw): bool =
+  # Unsynchronized
+  assert not chan.isNil
+
+  if chan.isClosed():
+    # Already closed
+    result = false
+  else:
+    chan.closed.store(true, moRelaxed)
+    result = true
+
+proc channel_open_spsc(chan: ChannelRaw): bool =
+  # Unsynchronized
+  assert not chan.isNil
+
+  if not chan.isClosed():
+    # Already open
+    result = false
+  else:
+    chan.closed.store(false, moRelaxed)
+    result = true
+
+# "Generic" dispatch
+# ----------------------------------------------------------------------------------
+
+const
+  send_fn = [
+    Mpmc: sendMpmc,
+    Mpsc: channel_send_mpsc,
+    Spsc: channel_send_spsc
+  ]
+
+  recv_fn = [
+    Mpmc: recvMpmc,
+    Mpsc: channel_recv_mpsc,
+    Spsc: channel_recv_spsc
+  ]
+
+  close_fn = [
+    Mpmc: channel_close_mpmc,
+    Mpsc: channel_close_mpsc,
+    Spsc: channel_close_spsc
+  ]
+
+  open_fn = [
+    Mpmc: channel_open_mpmc,
+    Mpsc: channel_open_mpsc,
+    Spsc: channel_open_spsc
+  ]
+
+proc channel_send(chan: ChannelRaw, data: sink pointer, size: int32, nonBlocking: bool): bool {.inline.} =
+  ## Send item to the channel (FIFO queue)
+  ## (Insert at last)
+  send_fn[chan.impl](chan, data, size, nonBlocking)
+
+proc channel_receive(chan: ChannelRaw, data: pointer, size: int32, nonBlocking: bool): bool {.inline.} =
+  ## Receive an item from the channel
+  ## (Remove the first item)
+  recv_fn[chan.impl](chan, data, size, nonBlocking)
+
+proc channel_close(chan: ChannelRaw): bool {.inline.} =
+  ## Close a channel
+  close_fn[chan.impl](chan)
+
+proc channel_open(chan: ChannelRaw): bool {.inline.} =
+  ## (Re)open a channel
+  open_fn[chan.impl](chan)
+
+# Public API
+# ----------------------------------------------------------------------------------
+
+type
+  Chan*[T] = object ## Typed channels
+    d: ChannelRaw
+
+proc `=`[T](dest: var Chan[T]; src: Chan[T]) {.error.}
+
+proc `=destroy`[T](c: var Chan[T]) =
+  if c.d.buffer != nil: freeChannel(c.d)
+
+proc channel_send[T](chan: Chan[T], data: T, size: int32, nonBlocking: bool): bool {.inline.} =
+  ## Send item to the channel (FIFO queue)
+  ## (Insert at last)
+  send_fn[chan.d.impl](chan.d, data.unsafeAddr, size, nonBlocking)
+
+proc channel_receive[T](chan: Chan[T], data: ptr T, size: int32, nonBlocking: bool): bool {.inline.} =
+  ## Receive an item from the channel
+  ## (Remove the first item)
+  recv_fn[chan.d.impl](chan.d, data, size, nonBlocking)
+
+func trySend*[T](c: Chan[T], src: sink Isolated[T]): bool {.inline.} =
+  ## Sends item to the channel(non blocking).
+  var data = src.extract
+  result = channel_send(c, data, int32 sizeof(data), true)
+  if result:
+    wasMoved(data)
```

Comment:
This shouldn't be required, `src.extract` has to empty the `src`. I think with `var Isolated` this problem disappears too.